clayer/pysertype.c

```diff
@@ -189,67 +123,38 @@ static ddspy_serdata_t *ddspy_serdata_new(const struct ddsi_sertype* type, enum
     new->key_populated = false;
     new->data_is_key = false;
     new->is_v2 = ((ddspy_sertype_t*)type)->is_v2_by_default;
-    memset((unsigned char*) &(new->hash), 0, 16);
 
     return new;
 }
 
-static void ddspy_serdata_calc_hash(ddspy_serdata_t* this)
+static void ddspy_serdata_populate_key(ddspy_serdata_t* this)
 {
-    if (this->is_v2) {
-        if (csertype(this)->v2_key_maxsize_bigger_16) {
-            ddsrt_md5_state_t md5st;
-            ddsrt_md5_init(&md5st);
-            ddsrt_md5_append(&md5st, (void*)(((char*)this->key) + 4), (unsigned int)this->key_size - 4);
-            ddsrt_md5_finish(&md5st, this->hash.value);
-        } else {
-            assert(this->key_size <= 20);
-            memset(this->hash.value, 0, 16);
-            memcpy(this->hash.value, ((char*)this->key) + 4, this->key_size - 4);
-        }
-        this->c_data.hash = ddsrt_mh3(this->key, this->key_size, 0) ^ this->c_data.type->serdata_basehash;
+    if (csertype(this)->keyless) {
+        this->key = NULL;
+        this->key_size = 0;
+        this->key_populated = true;
     } else {
-        if (csertype(this)->v0_key_maxsize_bigger_16) {
-            ddsrt_md5_state_t md5st;
-            ddsrt_md5_init(&md5st);
-            ddsrt_md5_append(&md5st, (void*)(((char*)this->key) + 4), (unsigned int)this->key_size - 4);
-            ddsrt_md5_finish(&md5st, this->hash.value);
+        const uint32_t xcdr_version = this->is_v2 ? DDSI_RTPS_CDR_ENC_VERSION_2 : DDSI_RTPS_CDR_ENC_VERSION_1;
+        void *cdr_hdr = this->data;
+        void *cdr_data = this->data + 4;
+
+        dds_ostream_t os;
+        dds_ostream_init(&os, &cdrstream_allocator, 0, xcdr_version);
+        dds_istream_t is;
+        dds_istream_init(&is, this->data_size - 4, cdr_data, xcdr_version);
+
+        if (dds_stream_extract_key_from_data(&is, &os, &cdrstream_allocator, &csertype(this)->cdrstream_desc)) {
+            this->key_size = os.m_index + 4;
+            this->key = dds_alloc(this->key_size);
+            memcpy(this->key, cdr_hdr, 4);
+            memcpy(this->key + 4, os.m_buffer, os.m_index);
+            this->key_populated = true;
         } else {
-            assert(this->key_size <= 20);
-            memset(this->hash.value, 0, 16);
-            memcpy(this->hash.value, ((char*)this->key) + 4, this->key_size - 4);
+            this->key_populated = false;
         }
```

Comment:
Isn't there a `dds_ostream_fini` call missing here?

Reply:
Yes, you're right, thanks! This one and the 2 below are fixed in [2a0c166](https://github.com/eclipse-cyclonedds/cyclonedds-python/pull/199/commits/2a0c1668ff3cfce73f6ea4c6dd3bad2b81eef9ea). I've also checked with address sanitizer if there are more leaks, but except for a lot of false positives on Python mallocs, these 3 were the only ones reported.